crypto/pkcs12/p12_crt.c

```diff
@@ -254,6 +254,7 @@ PKCS12_SAFEBAG *PKCS12_add_key_ex(STACK_OF(PKCS12_SAFEBAG) **pbags,
         PKCS8_PRIV_KEY_INFO_free(p8);
     } else
         bag = PKCS12_SAFEBAG_create0_p8inf(p8);
+    p8 = NULL;
 
     if (!bag)
         goto err;
```

Comment:
Something's not right here...
```suggestion
        PKCS8_PRIV_KEY_INFO_free(p8);
        p8 = NULL;
    } else
        bag = PKCS12_SAFEBAG_create0_p8inf(p8);

    if (!bag)
        goto err;

    p8 = NULL;
```

`PKCS12_SAFEBAG_create_pkcs8_encrypt_ex()` never holds onto `p8`, so, it is correctly freed immediately afterwards, but `p8` must also be set to `NULL` after.

However, `PKCS12_SAFEBAG_create_8_p8inf()` might hold onto `p8`, so, 'p8` must be set to NULL _only if_ `bag` does not return `NULL`. Otherwise, `p8` is leaked.

Reply:
I admit it's an ugly suggested change.

Reply:
why do you think that it is safe to assume that `PKCS12_SAFEBAG_create_pkcs8_encrypt_ex()` will always consume `p8` ?
what if an early error happens in `PKCS8_encrypt_ex()` or `PKCS12_SAFEBAG_create0_pkcs8()` ?